src/bindings/js.ts

```diff
@@ -1367,3 +1367,56 @@ function indentText(text: string, indentLevel: i32, sb: string[], butFirst: bool
     sb.push(text.substring(lineStart));
   }
 }
+
+export function liftRequiresExportRuntime(type: Type): bool {
+  if (!type.isInternalReference) return false;
+  let clazz = type.classReference;
+  if (!clazz) {
+    // functions lift as internref using __pin
+    assert(type.signatureReference);
+    return true;
+  }
+  let program = clazz.program;
+  // flat collections lift via memory copy
+  if (
+    clazz.extends(program.arrayBufferInstance.prototype) ||
+    clazz.extends(program.stringInstance.prototype) ||
+    clazz.extends(program.arrayBufferViewInstance.prototype)
+  ) {
+    return false;
+  }
+  // nested collections lift depending on element type
+  if (
+    clazz.extends(program.arrayPrototype) ||
+    clazz.extends(program.staticArrayPrototype)
+  ) {
+    return liftRequiresExportRuntime(clazz.getArrayValueType());
+  }
+  // complex objects lift as internref using __pin. plain objects may or may not
+  // involve the runtime: assume that they do to avoid potentially costly checks
```

Comment:
I'm wondering if lifting usually required only `__pin` but not `__new`. Could we avoid export full runtime and export only `__pin` / `__unpin` which probably will take much less size overhead than full exportRuntime?

Reply:
lifting only requires `__pin`, yeah. the size win here doesn't come from whether either of `__pin` or `__new` is exported, but that *both* are not even compiled. exists one, exists the other, basically +/- a few bytes.

Reply:
thinking a bit more about this, if some things were not exported, perhaps Binaryen could do a better job at DCE. it's not currently built for this, though, in that the runtime is supposed to be exported as a whole. I guess adding more here would be a follow-up PR.

Reply:
Hmm, yeah I agree.